src/libstore/build.cc

```diff
@@ -3689,162 +3835,314 @@ void DerivationGoal::registerOutputs()
+        } else {
+            /* content-addressed case */
+            DerivationOutputCAFloating outputHash = std::visit(overloaded {
+                [&](DerivationOutputInputAddressed doi) -> DerivationOutputCAFloating {
+                    // Enclosing `if` handles this case in other branch
+                    throw Error("ought to unreachable, handled in other branch");
                 },
                 [&](DerivationOutputCAFixed dof) {
-                    outputHash = DerivationOutputCAFloating {
+                    return DerivationOutputCAFloating {
                         .method = dof.hash.method,
                         .hashType = dof.hash.hash.type,
                     };
                 },
                 [&](DerivationOutputCAFloating dof) {
-                    outputHash = dof;
+                   return dof;
                 },
-            }, i.second.output);
-
+            }, output.output);
+            auto & st = outputStats.at(outputName);
             if (outputHash.method == FileIngestionMethod::Flat) {
                 /* The output path should be a regular file without execute permission. */
                 if (!S_ISREG(st.st_mode) || (st.st_mode & S_IXUSR) != 0)
                     throw BuildError(
                         "output path '%1%' should be a non-executable regular file "
                         "since recursive hashing is not enabled (outputHashMode=flat)",
-                        path);
+                        actualPath);
             }
-
-            /* Check the hash. In hash mode, move the path produced by
-               the derivation to its content-addressed location. */
-            Hash h2 = outputHash.method == FileIngestionMethod::Recursive
-                ? hashPath(outputHash.hashType, actualPath).first
-                : hashFile(outputHash.hashType, actualPath);
-
-            auto dest = worker.store.makeFixedOutputPath(outputHash.method, h2, i.second.path(worker.store, drv->name).name());
-
-            // true if either floating CA, or incorrect fixed hash.
-            bool needsMove = true;
-
-            if (auto p = std::get_if<DerivationOutputCAFixed>(& i.second.output)) {
-              Hash & h = p->hash.hash;
-              if (h != h2) {
-
-                /* Throw an error after registering the path as
-                   valid. */
-                worker.hashMismatch = true;
-                delayedException = std::make_exception_ptr(
-                    BuildError("hash mismatch in fixed-output derivation '%s':\n  wanted: %s\n  got:    %s",
-                        worker.store.printStorePath(dest),
-                        h.to_string(SRI, true),
-                        h2.to_string(SRI, true)));
-              } else {
-                  // matched the fixed hash, so no move needed.
-                  needsMove = false;
-              }
+            rewriteOutput();
+            /* FIXME optimize and deduplicate with addToStore */
```

Comment:
More than just `addToStore`, this is morally a duplicate of `CmdMakeContentAddressable::run` (modulo a couple of details like the fact that the original path isn't a valid path)

Reply:
Agreed. I just think those details, while conceptually insignificant, will make abstracting over this and `CmdMakeContentAddressable::run` harder. I am thinking start with making `addToStoreFromDump` references-aware, maybe then merge with `addTextToStore`, and only then re-compare with `CmdMakeContentAddressable::run`.

Reply:
Yep', makes sense. That's roughly what I'd done in b959baee6d51b44e51c8896c302ccd11fd0a458a (except that I don't use `addToStoreFromDump` because I wanted to keep the original pathInfo)

Reply:
Yeah, I have found the C++ non-trivial to deduplicate too. I think it's better to land the features and *then* choose the abstractions to make sure we don't abstract the wrong way----the plethora of `addToStore`-like functions kindof implies we've already been abstracting suboptimally, so I think a rethink should be very broad and systematic.